File: api/schemas/invite_user_to_org/invite_user_to_org.py
import os
import graphene

from graphql import GraphQLError
from notifications_python_client.notifications import NotificationsAPIClient

from app import logger
from db import db_session
from enums.languages import LanguageEnums
from enums.roles import RoleEnums
from functions.auth_functions import is_admin
from functions.auth_wrappers import require_token
from functions.input_validators import cleanse_input
from models import Organizations, User_affiliations, Users
from scalars.email_address import EmailAddress
from scalars.slug import Slug
from schemas.invite_user_to_org.send_added_to_org_email import (
    send_invite_to_org_notification_email,
)
from schemas.invite_user_to_org.send_invite_to_service_email import (
    send_invite_to_service_email,
)


class InviteUserToOrgInput(graphene.InputObjectType):
    """
    An input object containing all input fields for the InviteUserToOrg mutation
    """

    user_name = EmailAddress(
        required=True,
        description="User's email that you would like to invite to your org.",
    )
    requested_role = RoleEnums(
        required=True, description="The role which you would like this user to have."
    )
    org_slug = Slug(
        required=True, description="The organization you wish to invite the user to."
    )
    preferred_language = LanguageEnums(
        required=True,
        description="Select the language you would like the email to be sent in.",
    )


class InviteUserToOrg(graphene.Mutation):
    """
    This mutation allows admins and higher to invite users to any of their
    organizations, if the invited user does not have an account, they will be
    able to sign-up and be assigned to that organization in one mutation.
    """

    class Arguments:
        input = InviteUserToOrgInput(
            required=True,
            description="Input object containing fields for inviteUserToOrg mutation.",
        )

    status = graphene.String(description="Status of inviting user to organization.")

    @require_token
    def mutate(self, info, **kwargs):
        """
        This mutation does the work for the inviteUserToOrg mutation
        :param info: Request information
        :param kwargs: Various arguments from the user, and token information
        :return: A InviteUserToOrg object with a status field set as a string
        """
        # Get Arguments
        user_id = kwargs.get("user_id")
        user_roles = kwargs.get("user_roles")
        user_name = cleanse_input(kwargs.get("input", {}).get("user_name"))
        requested_role = cleanse_input(kwargs.get("input", {}).get("requested_role"))
        org_slug = cleanse_input(kwargs.get("input", {}).get("org_slug"))
        preferred_language = cleanse_input(
            kwargs.get("input", {}).get("preferred_language")
        )

        # Check to make sure user_name and requested_role are not none
        if (
            (user_name is not None)
            and (requested_role is not None)
            and (org_slug is not None)
        ):
            # Check to see if user exists
            user = Users.find_by_user_name(user_name=user_name)
            org = (
                db_session.query(Organizations)
                .filter(Organizations.slug == org_slug)
                .first()
            )
            if user is not None and org is not None:
                if is_admin(user_roles=user_roles, org_id=org.id) is True:
                    # Stop if permission is super_admin
                    if requested_role == "super_admin":
                        raise GraphQLError(
                            "Error, you cannot invite a user with a super admin role."
                        )

                    # Check to see if affiliation already exists
                    check_affiliation = (
                        db_session.query(User_affiliations)
                        .filter(User_affiliations.user_id == user.id)
                        .filter(User_affiliations.organization_id == org.id)
                        .first()
                    )
                    if check_affiliation is not None:
                        logger.warning(
                            f"User: {user_id}, attempted to add {user.id} to {org_slug}, but that user is already assigned to that organization."
                        )
                        raise GraphQLError(
                            f"Error, that user is already part of {org_slug}."
                        )

                    new_user_affilition = User_affiliations(
                        organization_id=org.id,
                        user_organization=org,
                        user_id=user_id,
                        user=user,
                        permission=requested_role,
                    )

                    try:
                        if (
                            send_invite_to_org_notification_email(
                                user=user,
                                org_name=org.name,
                                client=NotificationsAPIClient(
                                    api_key=os.getenv("NOTIFICATION_API_KEY"),
                                    base_url=os.getenv("NOTIFICATION_API_URL"),
                                ),
                            )
                            is True
                        ):
                            db_session.add(new_user_affilition)
                            db_session.commit()
                            logger.info(
                                f"User: {user_id} successfully added {user.id} to {org_slug}."
                            )
                            return InviteUserToOrg(
                                status="Successfully invited user to organization, and sent notification email."
                            )
                    except Exception as e:
                        logger.error(
                            f"User: {user_id} attempted to invite {user.id} to org, but a db error occurred when adding user: {str(e)}"
                        )
                        raise GraphQLError(
                            "Error, invite user to org failed, please try again."
                        )
                else:
                    logger.warning(
                        f"User: {user_id} attempted to invite user to {org_slug}, but does not have admin access."
                    )
                    raise GraphQLError(
                        "Error, you do not have access to invite users this organization."
                    )

            elif user is None and org is not None:
                # Send Email
                if (
                    send_invite_to_service_email(
                        user_name=user_name,
                        org=org,
                        preferred_language=preferred_language,
                        requested_role=requested_role,
                        client=NotificationsAPIClient(
                            api_key=os.getenv("NOTIFICATION_API_KEY"),
                            base_url=os.getenv("NOTIFICATION_API_URL"),
                        ),
                    )
                    is True
                ):
                    logger.info(
                        f"User: {user_id}, sent an invitation email to {user_name}, for the {org_slug} organization."
                    )
                    return InviteUserToOrg(
                        status="Successfully sent invitation to service, and organization email."
                    )

            elif org is None:
                logger.warning(
                    f"User: {user_id}, attempted to invite {user_name} to {org_slug}, but org does not exist."
                )
                raise GraphQLError(
                    "Error, invite user to org failed, please try again."
                )

        else:
            logging_str = ""
            if user_name is None:
                logging_str += " user_name"
            if requested_role is None:
                logging_str += " requested_role"
            if org_slug is None:
                logging_str += " org_slug"
            logger.warning(
                f"User: {user_id} attempted to invite a user but{logging_str} field(s) were missing"
            )

            raise GraphQLError("Error, invite user to org failed, please try again.")

File: api/src/affiliation/mutations/invite-user-to-org.js
import { GraphQLNonNull, GraphQLID } from 'graphql'
import { mutationWithClientMutationId, fromGlobalId } from 'graphql-relay'
import { GraphQLEmailAddress } from 'graphql-scalars'
import { t } from '@lingui/macro'

import { inviteUserToOrgUnion } from '../unions'
import { logActivity } from '../../audit-logs/mutations/log-activity'
import { InvitationRoleEnums } from '../../enums'

export const inviteUserToOrg = new mutationWithClientMutationId({
  name: 'InviteUserToOrg',
  description: `This mutation allows admins and higher to invite users to any of their
organizations, if the invited user does not have an account, they will be
able to sign-up and be assigned to that organization in one mutation.`,
  inputFields: () => ({
    userName: {
      type: new GraphQLNonNull(GraphQLEmailAddress),
      description: 'Users email that you would like to invite to your org.',
    },
    requestedRole: {
      type: new GraphQLNonNull(InvitationRoleEnums),
      description: 'The role which you would like this user to have.',
    },
    orgId: {
      type: new GraphQLNonNull(GraphQLID),
      description: 'The organization you wish to invite the user to.',
    },
  }),
  outputFields: () => ({
    result: {
      type: inviteUserToOrgUnion,
      description:
        '`InviteUserToOrgUnion` returning either a `InviteUserToOrgResult`, or `InviteUserToOrgError` object.',
      resolve: (payload) => payload,
    },
  }),
  mutateAndGetPayload: async (
    args,
    {
      i18n,
      query,
      request,
      collections,
      transaction,
      userKey,
      auth: { checkPermission, tokenize, userRequired, verifiedRequired, tfaRequired },
      loaders: { loadOrgByKey, loadUserByUserName },
      notify: { sendOrgInviteCreateAccount, sendOrgInviteEmail },
      validators: { cleanseInput },
    },
  ) => {
    const userName = cleanseInput(args.userName).toLowerCase()
    const requestedRole = cleanseInput(args.requestedRole)
    const { id: orgId } = fromGlobalId(cleanseInput(args.orgId))

    // Get requesting user
    const user = await userRequired()

    verifiedRequired({ user })
    tfaRequired({ user })

    // Make sure user is not inviting themselves
    if (user.userName === userName) {
      console.warn(`User: ${userKey} attempted to invite themselves to ${orgId}.`)
      return {
        _type: 'error',
        code: 400,
        description: i18n._(t`Unable to invite yourself to an org.`),
      }
    }

    // Check to see if requested org exists
    const org = await loadOrgByKey.load(orgId)

    if (typeof org === 'undefined') {
      console.warn(
        `User: ${userKey} attempted to invite user: ${userName} to ${orgId} however there is no org associated with that id.`,
      )
      return {
        _type: 'error',
        code: 400,
        description: i18n._(t`Unable to invite user to unknown organization.`),
      }
    }

    // Check to see requesting users permission to the org is
    const permission = await checkPermission({ orgId: org._id })

    // Only admins, owners, and super admins may invite users to an org
    // Only super admins may create owners and other super admins
    if (
      (['user', 'admin'].includes(requestedRole) && !['admin', 'owner', 'super_admin'].includes(permission)) ||
      (['super_admin', 'owner'].includes(requestedRole) && permission !== 'super_admin')
    ) {
      console.warn(
        `User: ${userKey} attempted to invite user: ${userName} to org: ${org._key} with role: ${requestedRole} but does not have permission to do so.`,
      )
      return {
        _type: 'error',
        code: 403,
        description: i18n._(t`Permission Denied: Please contact organization admin for help with user invitations.`),
      }
    }

    // Get org names to use in email
    let orgNamesCursor
    try {
      orgNamesCursor = await query`
        LET org = DOCUMENT(organizations, ${org._id})
        RETURN {
          "orgNameEN": org.orgDetails.en.name,
          "orgNameFR": org.orgDetails.fr.name,
        }
      `
    } catch (err) {
      console.error(
        `Database error occurred when user: ${userKey} attempted to invite user: ${userName} to org: ${org._key}. Error while creating cursor for retrieving organization names. error: ${err}`,
      )
      throw new Error(i18n._(t`Unable to invite user to organization. Please try again.`))
    }

    let orgNames
    try {
      orgNames = await orgNamesCursor.next()
    } catch (err) {
      console.error(
        `Cursor error occurred when user: ${userKey} attempted to invite user: ${userName} to org: ${org._key}. Error while retrieving organization names. error: ${err}`,
      )
      throw new Error(i18n._(t`Unable to invite user to organization. Please try again.`))
    }

    // Check to see if requested user exists
    const requestedUser = await loadUserByUserName.load(userName)

    // If there is not associated account with that username send invite to org with create account
    if (typeof requestedUser === 'undefined') {
      const token = tokenize({
        expiresIn: '3d',
        parameters: { userName, orgKey: org._key, requestedRole },
      })
      const createAccountLink = `https://${request.get('host')}/create-user/${token}`

      await sendOrgInviteCreateAccount({
        user: { userName: userName },
        orgNameEN: orgNames.orgNameEN,
        orgNameFR: orgNames.orgNameFR,
        createAccountLink,
      })

      console.info(`User: ${userKey} successfully invited user: ${userName} to the service, and org: ${org.slug}.`)
      await logActivity({
        transaction,
        collections,
        query,
        initiatedBy: {
          id: user._key,
          userName: user.userName,
          role: permission,
        },
        action: 'add',
        target: {
          resource: userName,
          organization: {
            id: org._key,
            name: org.name,
          }, // name of resource being acted upon
          resourceType: 'user', // user, org, domain
          updatedProperties: [{ name: 'role', oldValue: '', newValue: requestedRole }],
        },
      })

      return {
        _type: 'regular',
        status: i18n._(t`Successfully sent invitation to service, and organization email.`),
      }
    }

    // If account is found, check if already affiliated with org
    let affiliationCursor
    try {
      affiliationCursor = await query`
        WITH affiliations, organizations, users
        FOR v, e IN 1..1 INBOUND ${requestedUser._id} affiliations
          FILTER e._from == ${org._id}
          RETURN e
      `
    } catch (err) {
      console.error(
        `Database error occurred when user: ${userKey} attempted to invite user: ${requestedUser._key} to org: ${org.slug}, error: ${err}`,
      )
      return {
        _type: 'error',
        code: 500,
        description: i18n._(t`Unable to invite user to organization. Please try again.`),
      }
    }

    if (affiliationCursor.count > 0) {
      // If affiliation is found, return error
      console.warn(
        `User: ${userKey} attempted to invite user: ${requestedUser._key} to org: ${org.slug} however they are already affiliated with that org.`,
      )
      return {
        _type: 'error',
        code: 400,
        description: i18n._(t`Unable to invite user to organization. User is already affiliated with organization.`),
      }
    }

    // User is not affiliated with org, create affiliation

    // Setup Transaction
    const trx = await transaction(collections)

    // Create affiliation
    try {
      await trx.step(
        () =>
          query`
            WITH affiliations, organizations, users
            INSERT {
              _from: ${org._id},
              _to: ${requestedUser._id},
              permission: ${requestedRole},
            } INTO affiliations
          `,
      )
    } catch (err) {
      console.error(
        `Transaction step error occurred while user: ${userKey} attempted to invite user: ${requestedUser._key} to org: ${org.slug}, error: ${err}`,
      )
      await trx.abort()
      return {
        _type: 'error',
        code: 500,
        description: i18n._(t`Unable to invite user. Please try again.`),
      }
    }

    await sendOrgInviteEmail({
      user: requestedUser,
      orgNameEN: orgNames.orgNameEN,
      orgNameFR: orgNames.orgNameFR,
    })

    // Commit affiliation
    try {
      await trx.commit()
    } catch (err) {
      console.error(
        `Transaction commit error occurred while user: ${userKey} attempted to invite user: ${requestedUser._key} to org: ${org.slug}, error: ${err}`,
      )
      await trx.abort()
      return {
        _type: 'error',
        code: 500,
        description: i18n._(t`Unable to invite user. Please try again.`),
      }
    }

    console.info(`User: ${userKey} successfully invited user: ${requestedUser._key} to the org: ${org.slug}.`)
    await logActivity({
      transaction,
      collections,
      query,
      initiatedBy: {
        id: user._key,
        userName: user.userName,
        role: permission,
      },
      action: 'add',
      target: {
        resource: userName,
        organization: {
          id: org._key,
          name: org.name,
        }, // name of resource being acted upon
        updatedProperties: [{ name: 'role', oldValue: '', newValue: requestedRole }],
        resourceType: 'user', // user, org, domain
      },
    })

    return {
      _type: 'regular',
      status: i18n._(t`Successfully invited user to organization, and sent notification email.`),
    }
  },
})
